### ISSUE TITLE
LSTM checkpoint loading failure?

### ISSUE THREAD

**Issue by w3nhao** [jaem-seo/AI_tokamak_control#1]
Hi Jaemin, I use the codes and weights from your repo and when I run the scripts, it seems the LSTM model is not successfully loaded, however the simulation and visualization seem working just fine, would that be a problem?

For example when I run ```python rt_control_v2.py```, it raises warning like "W tensorflow/core/util/tensor_slice_reader.cc:95] Could not open /data/kstar/weights/lstm/v220505/best_model0: Data loss: not an sstable (bad magic number): perhaps your file is in a different file format and you need to use a different restore operator?"

Detail logs as below:
```bash
2025-01-03 15:10:18.056354: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcudart.so.11.0
start Xvnc
connect with: vncviewer localhost:5904
QStandardPaths: XDG_RUNTIME_DIR not set, defaulting to '/tmp/runtime-root'
2025-01-03 15:10:19.026399: I tensorflow/compiler/jit/xla_cpu_device.cc:41] Not creating XLA devices, tf_xla_enable_xla_devices not set
2025-01-03 15:10:19.027527: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcuda.so.1
2025-01-03 15:10:19.044094: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1720] Found device 0 with properties: 
pciBusID: 0000:34:00.0 name: NVIDIA RTX A6000 computeCapability: 8.6
coreClock: 1.8GHz coreCount: 84 deviceMemorySize: 47.54GiB deviceMemoryBandwidth: 715.34GiB/s
2025-01-03 15:10:19.044370: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1720] Found device 1 with properties: 
pciBusID: 0000:35:00.0 name: NVIDIA RTX A6000 computeCapability: 8.6
coreClock: 1.8GHz coreCount: 84 deviceMemorySize: 47.54GiB deviceMemoryBandwidth: 715.34GiB/s
2025-01-03 15:10:19.044626: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1720] Found device 2 with properties: 
pciBusID: 0000:36:00.0 name: NVIDIA RTX A6000 computeCapability: 8.6
coreClock: 1.8GHz coreCount: 84 deviceMemorySize: 47.54GiB deviceMemoryBandwidth: 715.34GiB/s
2025-01-03 15:10:19.044877: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1720] Found device 3 with properties: 
pciBusID: 0000:37:00.0 name: NVIDIA RTX A6000 computeCapability: 8.6
coreClock: 1.8GHz coreCount: 84 deviceMemorySize: 47.54GiB deviceMemoryBandwidth: 715.34GiB/s
2025-01-03 15:10:19.045128: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1720] Found device 4 with properties: 
pciBusID: 0000:9a:00.0 name: NVIDIA RTX A6000 computeCapability: 8.6
coreClock: 1.8GHz coreCount: 84 deviceMemorySize: 47.54GiB deviceMemoryBandwidth: 715.34GiB/s
2025-01-03 15:10:19.045379: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1720] Found device 5 with properties: 
pciBusID: 0000:9b:00.0 name: NVIDIA RTX A6000 computeCapability: 8.6
coreClock: 1.8GHz coreCount: 84 deviceMemorySize: 47.54GiB deviceMemoryBandwidth: 715.34GiB/s
2025-01-03 15:10:19.045633: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1720] Found device 6 with properties: 
pciBusID: 0000:9c:00.0 name: NVIDIA RTX A6000 computeCapability: 8.6
coreClock: 1.8GHz coreCount: 84 deviceMemorySize: 47.54GiB deviceMemoryBandwidth: 715.34GiB/s
2025-01-03 15:10:19.045885: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1720] Found device 7 with properties: 
pciBusID: 0000:9e:00.0 name: NVIDIA RTX A6000 computeCapability: 8.6
coreClock: 1.8GHz coreCount: 84 deviceMemorySize: 47.54GiB deviceMemoryBandwidth: 715.34GiB/s
2025-01-03 15:10:19.045906: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcudart.so.11.0
2025-01-03 15:10:19.050669: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcublas.so.11
2025-01-03 15:10:19.050718: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcublasLt.so.11
2025-01-03 15:10:19.052932: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcufft.so.10
2025-01-03 15:10:19.053222: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcurand.so.10
2025-01-03 15:10:19.053411: W tensorflow/stream_executor/platform/default/dso_loader.cc:60] Could not load dynamic library 'libcusolver.so.10'; dlerror: libcusolver.so.10: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /root/.mujoco/mujoco210/bin:/root/miniconda3/envs/rl_env/lib/python3.12/site-packages/nvidia/cusparse/lib:/usr/local/cuda-12.1/lib64:
2025-01-03 15:10:19.054382: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcusparse.so.11
2025-01-03 15:10:19.054471: W tensorflow/stream_executor/platform/default/dso_loader.cc:60] Could not load dynamic library 'libcudnn.so.8'; dlerror: libcudnn.so.8: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /root/.mujoco/mujoco210/bin:/root/miniconda3/envs/rl_env/lib/python3.12/site-packages/nvidia/cusparse/lib:/usr/local/cuda-12.1/lib64:
2025-01-03 15:10:19.054483: W tensorflow/core/common_runtime/gpu/gpu_device.cc:1757] Cannot dlopen some GPU libraries. Please make sure the missing libraries mentioned above are installed properly if you would like to use GPU. Follow the guide at https://www.tensorflow.org/install/gpu for how to download and setup the required libraries for your platform.
Skipping registering GPU devices...
2025-01-03 15:10:19.054716: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 AVX512F FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2025-01-03 15:10:19.061500: I tensorflow/compiler/jit/xla_gpu_device.cc:99] Not creating XLA devices, tf_xla_enable_xla_devices not set
2025-01-03 15:10:19.061526: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1261] Device interconnect StreamExecutor with strength 1 edge matrix:
2025-01-03 15:10:19.061531: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1267]      
2025-01-03 15:10:19.422028: W tensorflow/core/util/tensor_slice_reader.cc:95] Could not open /data/kstar/weights/lstm/v220505/best_model0: Data loss: not an sstable (bad magic number): perhaps your file is in a different file format and you need to use a different restore operator?
2025-01-03 15:10:19.728099: W tensorflow/core/util/tensor_slice_reader.cc:95] Could not open /data/kstar/weights/lstm/v220505/best_model1: Data loss: not an sstable (bad magic number): perhaps your file is in a different file format and you need to use a different restore operator?
2025-01-03 15:10:20.142372: W tensorflow/core/util/tensor_slice_reader.cc:95] Could not open /data/kstar/weights/lstm/v220505/best_model2: Data loss: not an sstable (bad magic number): perhaps your file is in a different file format and you need to use a different restore operator?
2025-01-03 15:10:20.442525: W tensorflow/core/util/tensor_slice_reader.cc:95] Could not open /data/kstar/weights/lstm/v220505/best_model3: Data loss: not an sstable (bad magic number): perhaps your file is in a different file format and you need to use a different restore operator?
2025-01-03 15:10:20.730442: W tensorflow/core/util/tensor_slice_reader.cc:95] Could not open /data/kstar/weights/lstm/v220505/best_model4: Data loss: not an sstable (bad magic number): perhaps your file is in a different file format and you need to use a different restore operator?
2025-01-03 15:10:21.011894: W tensorflow/core/util/tensor_slice_reader.cc:95] Could not open /data/kstar/weights/lstm/v220505/best_model5: Data loss: not an sstable (bad magic number): perhaps your file is in a different file format and you need to use a different restore operator?
2025-01-03 15:10:21.290600: W tensorflow/core/util/tensor_slice_reader.cc:95] Could not open /data/kstar/weights/lstm/v220505/best_model6: Data loss: not an sstable (bad magic number): perhaps your file is in a different file format and you need to use a different restore operator?
2025-01-03 15:10:21.571667: W tensorflow/core/util/tensor_slice_reader.cc:95] Could not open /data/kstar/weights/lstm/v220505/best_model7: Data loss: not an sstable (bad magic number): perhaps your file is in a different file format and you need to use a different restore operator?
2025-01-03 15:10:21.851249: W tensorflow/core/util/tensor_slice_reader.cc:95] Could not open /data/kstar/weights/lstm/v220505/best_model8: Data loss: not an sstable (bad magic number): perhaps your file is in a different file format and you need to use a different restore operator?
2025-01-03 15:10:22.372647: W tensorflow/core/util/tensor_slice_reader.cc:95] Could not open /data/kstar/weights/lstm/v220505/best_model9: Data loss: not an sstable (bad magic number): perhaps your file is in a different file format and you need to use a different restore operator?
2025-01-03 15:10:23.039957: I tensorflow/compiler/mlir/mlir_graph_optimization_pass.cc:116] None of the MLIR optimization passes are enabled (registered 2)
2025-01-03 15:10:23.040405: I tensorflow/core/platform/profile_utils/cpu_utils.cc:112] CPU Frequency: 2600000000 Hz
WARNING:tensorflow:5 out of the last 5 calls to <function Model.make_predict_function.<locals>.predict_function at 0x7f49d2c2e310> triggered tf.function retracing. Tracing is expensive and the excessive number of tracings could be due to (1) creating @tf.function repeatedly in a loop, (2) passing tensors with different shapes, (3) passing Python objects instead of tensors. For (1), please define your @tf.function outside of the loop. For (2), @tf.function has experimental_relax_shapes=True option that relaxes argument shapes that can avoid unnecessary retracing. For (3), please refer to https://www.tensorflow.org/guide/function#controlling_retracing and https://www.tensorflow.org/api_docs/python/tf/function for  more details.
WARNING:tensorflow:6 out of the last 6 calls to <function Model.make_predict_function.<locals>.predict_function at 0x7f49d2c2e670> triggered tf.function retracing. Tracing is expensive and the excessive number of tracings could be due to (1) creating @tf.function repeatedly in a loop, (2) passing tensors with different shapes, (3) passing Python objects instead of tensors. For (1), please define your @tf.function outside of the loop. For (2), @tf.function has experimental_relax_shapes=True option that relaxes argument shapes that can avoid unnecessary retracing. For (3), please refer to https://www.tensorflow.org/guide/function#controlling_retracing and https://www.tensorflow.org/api_docs/python/tf/function for  more details.
WARNING:tensorflow:7 out of the last 11 calls to <function Model.make_predict_function.<locals>.predict_function at 0x7f49d2c6d310> triggered tf.function retracing. Tracing is expensive and the excessive number of tracings could be due to (1) creating @tf.function repeatedly in a loop, (2) passing tensors with different shapes, (3) passing Python objects instead of tensors. For (1), please define your @tf.function outside of the loop. For (2), @tf.function has experimental_relax_shapes=True option that relaxes argument shapes that can avoid unnecessary retracing. For (3), please refer to https://www.tensorflow.org/guide/function#controlling_retracing and https://www.tensorflow.org/api_docs/python/tf/function for  more details.
```

**Comment by simbot912** [comment]
This is a common TensorFlow checkpoint issue. A few things to check:

1. **File path mismatch** — The error mentions `/data/kstar/weights/lstm/v220505/best_model0` which looks like an absolute path hardcoded in the model. If the weights aren't at that exact path, it fails. Try checking the config file or code for hardcoded paths.

2. **TensorFlow version compatibility** — Checkpoints saved with older TF versions (1.x) may not load cleanly in TF 2.x. Try adding `tf.compat.v1` prefix or using `tf.train.load_checkpoint()` to inspect the variables.

3. **Corrupted download** — If you cloned from GitHub, large checkpoint files might not have been properly transferred (Git LFS issues). Verify the file sizes match what's documented.

4. **Quick workaround** — If the simulation works despite the warning, the LSTM might be falling back to random initialization. You could retrain or find the correct checkpoint path.

Would recommend checking if the weights directory has the proper `.index` and `.data` files that TensorFlow expects.
